File: frontend/src/services/stockService.js
import httpClient from './validators/api/httpClient';
import { exportStockXLSX } from "./xlsxExporter";
import { logger } from "../utils/logger";

const FILIAL_OPTIONS = ["Porto Alegre", "Joinville", "São Paulo"];

const normalizeFilial = (value) => {
    const text = String(value || "").trim();
    if (!text) return "";
    return FILIAL_OPTIONS.includes(text) ? text : "";
};

// Adicionado o parâmetro 'index' para criar uma chave única
const mapStockToFrontend = (item, index) => {
    let rawSupplier = item.primary_supplier || item.supplier_name || item.fornecedor || item.supplier || item.suppliers?.name;
    let supplierStr = String(rawSupplier || "").trim();

    if (supplierStr.includes("Padrão") || supplierStr.includes("Default") || supplierStr === "null" || supplierStr === "undefined") {
        supplierStr = "";
    }

    const baseId = item.id || item.sku_id || "gen";
    const uniqueReactId = `${baseId}-${index}-${Math.random().toString(36).substr(2, 5)}`;

    return {
        id: uniqueReactId, // ID Único para o DataGrid renderizar sem erros
        real_sku_id: item.sku_id || item.id, // O ID real guardado para quando for criar a encomenda (pedido)
        codigo: item.sku_code || item.codigo || item.tb_skus?.codigo || "S/C",
        item: item.name || item.item || item.tb_skus?.nome_produto || "Item sem nome",
        categoria: item.category || item.categoria || "Geral",
        unidades: item.stock_quantity || item.unidades || item.estoque || item.estoque_soma || 0,
        fornecedor: supplierStr, 
        filial: normalizeFilial(item.filial || item.branch_name || item.branch || item.filial_nome),
        dias_cobertura: item.dias_cobertura, 
        valor: item.unit_price || item.valor || item.preco || item.preco_custo || 0,
        porto_alegre: item.porto_alegre || item.estoque_poa || 0,
        joinville: item.joinville || item.estoque_jv || 0,  
        sao_paulo: item.sao_paulo || item.estoque_sp || 0,
        rop: Math.ceil(item.rop || 0),
        qtd_sugerida: Math.ceil(item.qtd_sugerida || 0),
        leadtime: item.leadtime || 0,
        _raw: item
    };
};

export const getStockData = async (filial, fornecedor, status) => {
    try {
        const args = typeof filial === 'object' ? filial : { filial, fornecedor, status };

        const params = new URLSearchParams();
        if (args.filial && args.filial !== "Todos") params.append('filial', args.filial);
        if (args.fornecedor && args.fornecedor !== "Todos") params.append('fornecedor', args.fornecedor);
        if (args.status && args.status !== "Todos") params.append('status', args.status);

        const queryString = params.toString();
        const endpoint = queryString ? `/stock?${queryString}` : '/stock';
        
        const response = await httpClient.get(endpoint);
        const dataList = Array.isArray(response) ? response : (response.data || []);
        
        // Passa o item e o índice (index) para a função de mapeamento
        return dataList.map((item, index) => mapStockToFrontend(item, index));

    } catch (error) {
        logger.error("Erro ao carregar estoque:", error);
        return [];
    }
};

export const getSuppliers = async () => {
    try {
        let data = [];
        try {
            const res = await httpClient.get('/suppliers');
            data = Array.isArray(res) ? res : (res.data || []);
        } catch (e) {
            const resStock = await httpClient.get('/stock');
            data = Array.isArray(resStock) ? resStock : (resStock.data || []);
        }

        if (data.length > 0 && typeof data[0] === 'string') {
            return data;
        }

        const allSuppliers = data.map(item => {
            const val = item.name || item.supplier_name || item.primary_supplier || item.fornecedor || item.suppliers?.name;
            return String(val || "").trim();
        });
        
        return [...new Set(allSuppliers)]
            .filter(s => s && s !== "null" && s !== "undefined" && !s.includes("Padrão"))
            .sort();
            
    } catch (error) {
        logger.error("Erro ao buscar fornecedores:", error);
        return [];
    }
};

export const createOrderBatch = async (items) => {
    const safeItems = Array.isArray(items) ? items : [];
    const payload = { items: safeItems };

    try {
        const response = await httpClient.post('/orders/batch', payload);
        return response.data || response;
    } catch (error) {
        throw error;
    }
};

export const importStockFromFile = async (file) => {
    const formData = new FormData();
    formData.append("file", file);
    
    const API_URL = import.meta.env.VITE_API_URL || "http://localhost:8000";

    const response = await fetch(`${API_URL}/stock/import`, {
        method: "POST",
        body: formData,
        // ESTA É A LINHA QUE RESOLVE O 401: Envia o cookie de login para o backend
        credentials: "include",
    });

    const data = await response.json().catch(() => ({}));
    if (!response.ok) {
        throw new Error(data.detail || data.message || "Erro na requisição de importação");
    }

    return data;
};

export const exportStockData = async (data) => {
    if (!data || data.length === 0) throw new Error("Nenhum dado fornecido para exportação.");
    exportStockXLSX(data);
    return Promise.resolve({ message: "Dados do estoque exportados em Excel e download iniciado." });
};

File: backend/app/services/stock_service.py
import math
import logging
from typing import Optional, List
from app.repositories.stock_repository import StockRepository

logger = logging.getLogger(__name__)

class StockService:
    def __init__(self, stock_repository: Optional[StockRepository] = None):
        self.stock_repo = stock_repository or StockRepository()

    def get_stock(
        self,
        filial: Optional[str] = None,
        fornecedor: Optional[str] = None,
        status: Optional[str] = None,
        limit: Optional[int] = None,
        current_user: Optional[dict] = None,
        unidades_pendentes: Optional[int] = None,
    ) -> list:
        # Pega dados brutos da view via repositorio
        logger.info(
            "Buscando estoque - filial: %s - fornecedor: %s - limit: %s",
            filial, fornecedor, limit,
        )

        try:
            raw_data = self.stock_repo.get_stock_analysis(
                filial=filial,
                fornecedor=fornecedor,
                limit=limit,
                unidades_pendentes=unidades_pendentes,
            )
        except Exception:
            logger.exception(
                "Erro ao buscar análise de estoque - filial: %s - fornecedor: %s",
                filial,
                fornecedor,
            )
            raise
        if not raw_data:
            logger.info("Nenhum registro de estoque encontrado")
            return []
        
        filtered_data = []
        for item in raw_data:
            estoque_poa = item.get("estoque_poa") or 0
            estoque_jv = item.get("estoque_jv") or 0
            estoque_sp = item.get("estoque_sp") or 0

            if not filial or filial == "Todos":
                filtered_data.append(item)
            elif filial == "Porto Alegre" and float(estoque_poa) > 0:
                filtered_data.append(item)
            elif filial == "Joinville" and float(estoque_jv) > 0:
                filtered_data.append(item)
            elif (filial == "São Paulo" or "Paulo" in filial) and float(estoque_sp) > 0:
                filtered_data.append(item)

        result = []

        for item in raw_data:    
            try:      
                unique_id = str(item.get("sku_id"))
                
                def safe_float(val, default_to_zero=True):
                    if val is None:
                        return 0.0 if default_to_zero else None
                    try:
                        return float(val)
                    except (ValueError, TypeError):
                        return 0.0 if default_to_zero else None

                sugerida = safe_float(item.get("quantidade_sugerida_compra"))
                sugerida_projetada = safe_float(
                    item.get("quantidade_sugerida_compra_projetada", item.get("quantidade_sugerida_compra_v2"))
                )
                rop = safe_float(item.get("rop"))
                stock = safe_float(item.get("estoque_atual"))
                dias_cobertura = safe_float(item.get("dias_cobertura"), default_to_zero=False)
                unidades_pendentes = safe_float(item.get("unidades_pendentes"))
                estoque_projetado = safe_float(item.get("estoque_projetado"))
                dias_cobertura_projetado = safe_float(item.get("dias_cobertura_projetado"), default_to_zero=False)
                
                mapped_item = {
                    "id": unique_id,
                    "sku_id": item.get("sku_id"),
                    "codigo": item.get("codigo", "S/C"),
                    "item": item.get("nome_produto", "Item sem nome"),
                    "categoria": item.get("classificacao", "Geral"),
                    "unidades": int(stock),
                    "valor": 0, 
                    "fornecedor": item.get("fornecedor", "Não informado"),
                    "dias_cobertura": dias_cobertura if dias_cobertura is None else round(dias_cobertura, 2),
                    "porto_alegre": int(safe_float(item.get("estoque_poa"))),
                    "joinville": int(safe_float(item.get("estoque_jv"))),
                    "sao_paulo": int(safe_float(item.get("estoque_sp"))),
                    "filial": filial if (filial and filial != "Todos") else "Todos",
                    "rop": math.ceil(rop),
                    "qtd_sugerida": math.ceil(sugerida),
                    "leadtime": int(safe_float(item.get("leadtime_utilizado_dias"))),
                    "demanda_mensal": safe_float(item.get("demanda_mensal_media")),
                    "unidades_pendentes": int(unidades_pendentes),
                    "estoque_projetado": int(estoque_projetado),
                    "dias_cobertura_projetado": (
                        dias_cobertura_projetado if dias_cobertura_projetado is None else round(dias_cobertura_projetado, 2)
                    ),
                    "quantidade_sugerida_compra_projetada": math.ceil(sugerida_projetada)
                }
                result.append(mapped_item)

            except Exception:
                logger.exception(
                    "Erro ao processar item de estoque - sku_id: %s",
                    item.get("sku_id") if isinstance(item, dict) else "unknown",
                )
                continue
    
        return result